server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java

```diff
@@ -0,0 +1,207 @@
+     * and then unblocking the master service again so everything processes in a single batch.
+     *
+     * But closing and adding blocks are both two-step processes where the second step is itself another cluster state update task, and we
+     * want to make sure that those batch up correctly, too. So there we run the same trick as with opens, but twice: we block the
+     * master service in order to batch up the initial tasks, then block it again, so that when the first tasks enqueue the second tasks
+     * they're doing so into a blocked master service, finally the master service is unblocked once more so that the second tasks process
+     * together in a batch.
+     */
+
+    public void testBatchOpenIndices() throws Exception {
+        final var masterService = getInstanceFromNode(ClusterService.class).getMasterService();
+
+        // create some indices
+        createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
+        createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
+        createIndex("test-3", client().admin().indices().prepareCreate("test-3"));
+        ensureGreen("test-1", "test-2", "test-3");
+
+        final var block1 = blockMasterService(masterService);
+        block1.call(); // wait for block
+
+        // fire off some opens
+        final var future1 = client().admin().indices().prepareOpen("test-1").execute();
+        final var future2 = client().admin().indices().prepareOpen("test-2", "test-3").execute();
+
+        // check the queue for the open-indices tasks
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
+
+        block1.call(); // release block
+
+        // assert that the requests were acknowledged
+        assertAcked(future1.get());
+        assertAcked(future2.get());
+    }
+
+    public void testBatchCloseIndices() throws Exception {
+        final var masterService = getInstanceFromNode(ClusterService.class).getMasterService();
+
+        // create some indices
+        createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
+        createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
+        createIndex("test-3", client().admin().indices().prepareCreate("test-3"));
+        ensureGreen("test-1", "test-2", "test-3");
+
+        final var block1 = blockMasterService(masterService);
+        block1.call(); // wait for block
+
+        // fire off some closes
+        final var future1 = client().admin().indices().prepareClose("test-1").execute();
+        final var future2 = client().admin().indices().prepareClose("test-2", "test-3").execute();
+
+        // check the queue for the first close tasks (the add-block-index-to-close tasks)
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
```

Comment:
Similarly, this should hold immediately.

Reply:
4137dcaf9080b3df9517a1f348f91122ab9d625c